File: src/lib/registries/pypi.ts
import type { ResolvedPackage } from "../../types.js";

const PYPI_API = "https://pypi.org/pypi";

interface PyPIRelease {
  upload_time: string;
  yanked: boolean;
}

interface PyPIResponse {
  info: {
    name: string;
    version: string;
    home_page?: string;
    project_urls?: Record<string, string>;
    project_url?: string;
  };
  releases: Record<string, PyPIRelease[]>;
}

/**
 * Parse a PyPI package specifier like "requests==2.31.0" into name and version
 */
export function parsePyPISpec(spec: string): {
  name: string;
  version?: string;
} {
  // Handle version specifiers: requests==2.31.0 or requests>=2.31.0
  const eqMatch = spec.match(/^([^=<>!~]+)==(.+)$/);
  if (eqMatch) {
    return { name: eqMatch[1].trim(), version: eqMatch[2].trim() };
  }

  // Handle @ version specifier: requests@2.31.0
  const atIndex = spec.lastIndexOf("@");
  if (atIndex > 0) {
    return {
      name: spec.slice(0, atIndex).trim(),
      version: spec.slice(atIndex + 1).trim(),
    };
  }

  return { name: spec.trim() };
}

/**
 * Fetch package metadata from PyPI
 */
async function fetchPyPIPackageInfo(
  packageName: string,
  version?: string,
): Promise<PyPIResponse> {
  const url = version
    ? `${PYPI_API}/${packageName}/${version}/json`
    : `${PYPI_API}/${packageName}/json`;

  const response = await fetch(url, {
    headers: {
      Accept: "application/json",
    },
  });

  if (!response.ok) {
    if (response.status === 404) {
      throw new Error(`Package "${packageName}" not found on PyPI`);
    }
    throw new Error(
      `Failed to fetch package info: ${response.status} ${response.statusText}`,
    );
  }

  return response.json() as Promise<PyPIResponse>;
}

/**
 * Extract repository URL from PyPI package metadata
 */
function extractRepoUrl(info: PyPIResponse["info"]): string | null {
  // Check project_urls for common repository keys
  const projectUrls = info.project_urls || {};
  const repoKeys = [
    "Source",
    "Source Code",
    "Repository",
    "GitHub",
    "Code",
    "Homepage",
  ];

  for (const key of repoKeys) {
    const url = projectUrls[key];
    if (url && isGitRepoUrl(url)) {
      return normalizeRepoUrl(url);
    }
  }

  // Fall back to home_page if it's a git repo
  if (info.home_page && isGitRepoUrl(info.home_page)) {
    return normalizeRepoUrl(info.home_page);
  }

  // Check all project_urls for any git repo URL
  for (const url of Object.values(projectUrls)) {
    if (isGitRepoUrl(url)) {
      return normalizeRepoUrl(url);
    }
  }

  return null;
}

function isGitRepoUrl(url: string): boolean {
  return (
    url.includes("github.com") ||
    url.includes("gitlab.com") ||
    url.includes("bitbucket.org")
  );
}

function normalizeRepoUrl(url: string): string {
  // Remove trailing slashes and common suffixes
  return url
    .replace(/\/+$/, "")
    .replace(/\.git$/, "")
    .replace(/\/tree\/.*$/, "")
    .replace(/\/blob\/.*$/, "");
}

/**
 * Get available versions sorted by release date
 */
function getVersions(releases: PyPIResponse["releases"]): string[] {
  return Object.entries(releases)
    .filter(([, files]) => files.length > 0 && !files[0].yanked)
    .sort((a, b) => {
      const timeA = a[1][0]?.upload_time || "";
      const timeB = b[1][0]?.upload_time || "";
      return timeB.localeCompare(timeA);
    })
    .map(([version]) => version);
}

/**
 * Resolve a PyPI package to its repository information
 */
export async function resolvePyPIPackage(
  packageName: string,
  version?: string,
): Promise<ResolvedPackage> {
  const info = await fetchPyPIPackageInfo(packageName, version);

  const resolvedVersion = info.info.version;

  // Get available versions if we need them for error message
  let availableVersions: string[] = [];
  if (!version) {
    const fullInfo = await fetchPyPIPackageInfo(packageName);
    availableVersions = getVersions(fullInfo.releases).slice(0, 5);
  }

  const repoUrl = extractRepoUrl(info.info);

  if (!repoUrl) {
    throw new Error(
      `No repository URL found for "${packageName}@${resolvedVersion}". ` +
        `This package may not have its source published.` +
        (availableVersions.length > 0
          ? ` Recent versions: ${availableVersions.join(", ")}`
          : ""),
    );
  }

  // Python packages commonly use v1.2.3 or 1.2.3 as tags
  const gitTag = `v${resolvedVersion}`;

  return {
    registry: "pypi",
    name: packageName,
    version: resolvedVersion,
    repoUrl,
    gitTag,
  };
}

File: packages/opensrc/cli/src/lib/registries/pypi.rs
use super::{Registry, ResolvedPackage};
use serde::Deserialize;
use std::collections::HashMap;

const PYPI_API: &str = "https://pypi.org/pypi";

#[derive(Deserialize)]
struct PyPIInfo {
    version: String,
    home_page: Option<String>,
    project_urls: Option<HashMap<String, String>>,
}

#[derive(Deserialize)]
struct PyPIResponse {
    info: PyPIInfo,
}

pub fn parse_pypi_spec(spec: &str) -> (String, Option<String>) {
    // requests==2.31.0
    let re = regex::Regex::new(r"^([^=<>!~]+)==(.+)$").unwrap();
    if let Some(caps) = re.captures(spec) {
        return (caps[1].trim().to_string(), Some(caps[2].trim().to_string()));
    }

    // requests@2.31.0
    if let Some(at_idx) = spec.rfind('@') {
        if at_idx > 0 {
            return (
                spec[..at_idx].trim().to_string(),
                Some(spec[at_idx + 1..].trim().to_string()),
            );
        }
    }

    (spec.trim().to_string(), None)
}

fn fetch_pypi_info(
    name: &str,
    version: Option<&str>,
) -> Result<PyPIResponse, Box<dyn std::error::Error>> {
    let url = match version {
        Some(v) => format!("{PYPI_API}/{name}/{v}/json"),
        None => format!("{PYPI_API}/{name}/json"),
    };

    let client = reqwest::blocking::Client::new();
    let resp = client
        .get(&url)
        .header("Accept", "application/json")
        .send()?;

    if resp.status() == reqwest::StatusCode::NOT_FOUND {
        return Err(format!("Package \"{name}\" not found on PyPI").into());
    }
    if !resp.status().is_success() {
        return Err(format!("Failed to fetch package info: {}", resp.status()).into());
    }

    Ok(resp.json()?)
}

fn is_git_repo_url(url: &str) -> bool {
    url.contains("github.com") || url.contains("gitlab.com") || url.contains("bitbucket.org")
}

fn normalize_repo_url(url: &str) -> String {
    url.trim_end_matches('/')
        .trim_end_matches(".git")
        .split("/tree/")
        .next()
        .unwrap_or(url)
        .split("/blob/")
        .next()
        .unwrap_or(url)
        .to_string()
}

fn extract_repo_url(info: &PyPIInfo) -> Option<String> {
    let repo_keys = [
        "Source",
        "Source Code",
        "Repository",
        "GitHub",
        "Code",
        "Homepage",
    ];

    if let Some(urls) = &info.project_urls {
        for key in &repo_keys {
            if let Some(url) = urls.get(*key) {
                if is_git_repo_url(url) {
                    return Some(normalize_repo_url(url));
                }
            }
        }

        if let Some(hp) = &info.home_page {
            if is_git_repo_url(hp) {
                return Some(normalize_repo_url(hp));
            }
        }

        for url in urls.values() {
            if is_git_repo_url(url) {
                return Some(normalize_repo_url(url));
            }
        }
    }

    if let Some(hp) = &info.home_page {
        if is_git_repo_url(hp) {
            return Some(normalize_repo_url(hp));
        }
    }

    None
}

pub fn resolve_pypi_package(
    name: &str,
    version: Option<&str>,
) -> Result<ResolvedPackage, Box<dyn std::error::Error>> {
    let info = fetch_pypi_info(name, version)?;
    let resolved_version = info.info.version.clone();

    let repo_url = extract_repo_url(&info.info).ok_or_else(|| {
        format!(
            "No repository URL found for \"{name}@{resolved_version}\". \
             This package may not have its source published."
        )
    })?;

    let git_tag = format!("v{resolved_version}");

    Ok(ResolvedPackage {
        registry: Registry::PyPI,
        name: name.to_string(),
        version: resolved_version,
        repo_url,
        repo_directory: None,
        git_tag,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_pypi_spec_simple() {
        let (name, version) = parse_pypi_spec("requests");
        assert_eq!(name, "requests");
        assert_eq!(version, None);
    }

    #[test]
    fn test_parse_pypi_spec_double_eq() {
        let (name, version) = parse_pypi_spec("requests==2.31.0");
        assert_eq!(name, "requests");
        assert_eq!(version, Some("2.31.0".into()));
    }

    #[test]
    fn test_parse_pypi_spec_at() {
        let (name, version) = parse_pypi_spec("requests@2.31.0");
        assert_eq!(name, "requests");
        assert_eq!(version, Some("2.31.0".into()));
    }
}
